### ISSUE TITLE
Provide extension API to exclude ports from forwarding

### ISSUE THREAD

**Issue by alexr00** [microsoft/vscode#115616]
From @weinand:
Today the tunneling service blindly forwards all communication ports.
This includes ports that are used for debugging (even if our remote debugging feature does not need these ports to be forwarded).
This is confusing for users because they see ports that they are not really interested in.

I propose to add extension API so that individual ports or port ranges can be excluded from forwarding.
Debug extensions could use this API.

**Comment by alexr00** [comment]
We currently have a setting that lets user configure "ignored" ports:
```json
	"remote.portsAttributes": {
		"35000-38000": {
			"label": "My server",
			"onAutoForward": "ignore",
		}
	}
```

This setting also allows setting other properties of the port, such as the label.

Proposal:
```typescript

	export enum PortAutoForwardAction {
		Notify,
		OpenBrowser,
		Silent,
		Ignore
	}

	export interface PortAttributes {
		port: number;
		autoForwardAction: PortAutoForwardAction
	}

	export interface PortAttributesProvider {
		providePortAttributes(port: number[], token: CancellationToken): ProviderResult<PortAttributes[]>;
	}

	export namespace workspace {
		export function registerPortAttributesProvider(portRange: [number, number], provider: PortAttributesProvider): Disposable;
	}
```

**Comment by alexr00** [comment]
I discussed with @weinand and there are several places unwanted ports can come from:

a) Debugees. These are started by debugger extensions, and the debug internals of VS Code know nothing about them.
b) Debug extensions setting up a server for communication with the debuggee. This is what JS debug does.
c) Debug adapters. The debug internals of VS Code actually do know about this, but this isn't where we're seeing most unwanted ports come from.
d) Language servers. I haven't see reports of this, but it could absolutely happen.
e) Extensions in general. Some of these servers could be user facing, and some may not be. We don't have data on this.

Cases (a) and (b) above are the most common cases that we want to solve for. We discussed excluding all processes that come from the extension host from auto forwarded, but that does not solve case (a).

@connor4312 since JS debug would be the first client of any API I'd love to get your input here. The above `PortsAttributesProvider` API is very ports centric. We could instead have a `ProcessDetailProvider`, which returns something that indicates whether a process is intended to be user-facing.

**Comment by connor4312** [comment]
The API seems generally sensible to me.

While a consumer could pass it as `[0, 2 ** 16]`, it may be more ergonomic for the `portRange` to be optional: for debugging I always request port `0` and let the OS choose a free port to use. I imagine that other consumers frequently do this too.

Particularly in this case, providers will often get called with ports they know nothing about. It looks like `providePortAttributes` can take a bulk list of ports, but it can only return information for none of them (undefined) or all of them. If my debug port is mixed in a list of other ports, I can't say to ignore that one and fall back on the others. So maybe it's better to have it take a single port and return a ProviderResult<PortAttributes> for the one port.

**Comment by weinand** [comment]
@connor4312 you said
> "I always request port 0 and let the OS choose a free port to use."

When does your DA or extension know what the free port is?
I'm asking because the port forwarding service will call `PortAttributesProvider.providePortAttributes` as soon as it detects new ports through scanning and it would be a problem if your DA would not know at that time that the given port is a debug port.

---
and you said:
> "providers will often get called with ports they know nothing about" but there is no way "to ignore that one". 

The return type of `PortAttributesProvider.providePortAttributes` could be an array of `PortAttributes` or `undefined`, that is `ProviderResult<(PortAttributes | undefined)[]>`.

**Comment by connor4312** [comment]
> it would be a problem if your DA would not know at that time that the given port is a debug port.

Yea, that's correct. I would need to wait a few moments so that any port resolutions can happen. Or if this becomes problematic I could restrict it to a specific port range, then manually choose a free port and only have the waiting logic for ports in that range

> could be an array of 

Yep

**Comment by alexr00** [comment]
That should still be ok then. Even if you have to wait a few moments, you already know that you need to wait a few moments. So if the call comes from VS Code to your extension asking your extension to provide port information, you can wait the time you need to know the port number before returning the port information.

**Comment by alexr00** [comment]
New proposal:
```ts
	export interface ProcessInformation {
		isUserFacing: boolean;
	}

	export interface ProcessInformationProvider {
		provideProcessInformation(process: { pid: number, port: number }, token: CancellationToken): ProviderResult<ProcessInformation>;
	}

	export namespace workspace {
		export function registerProcessInformationProvider(selector: { portRange?: [number, number], processRange?: [number, number] }, processInformationProvider: ProcessInformationProvider): Disposable;
	}
```

**Comment by alexr00** [comment]
@connor4312 the API includes the pid, and I imagine that you'll know the pid almost immediately. Is that correct?

**Comment by connor4312** [comment]
In the auto attach / debug terminal case, not any earlier than I will the port: the debugee still needs to connect back to the extension and tell us it exists before we'll know its port. But for less esoteric cases passing the pid is good, I'm sure other consumers will appreciate that.

API looks good now. I'm not sure how useful the process range is -- as far as I know all OSes assign process IDs with an increment counter, so an extension would not know the process ID of whatever they're interested in ahead of time.

**Comment by connor4312** [comment]
I think using a provider in combination with a ports range is ideal, in order to avoid accidentally excluding user programs. Then the delay in that case would be less problematic: in the high likelihood that the queried port is actually a debugged program, it's not going to be shown to the user anyway.

**Comment by alexr00** [comment]
My (and @jrieken's) concern is that this a lot of API to add to support a very narrow scenario: don't auto forward debugee's debug ports. This API doesn't support any other scenario. Would you be willing to try the settings approach?

**Comment by jrieken** [comment]
Yes, the concern is that port forwarding, by its nature, forwards many unwanted ports. Ports from processes I am not interested in, ports that don't speak http, ports I cannot connect to etc. The screen shot below is what I see when I connect to my raspberry pi which run jupyter-lab. There are 7 ports forwarded (prominently announced via the status bar and notification) and I am only interested in one of them (iff at all)

<img width="1136" alt="image" src="https://user-images.githubusercontent.com/1794099/107756101-00868c00-6d24-11eb-9b27-79cdffd38cd6.png">

I know that debugging ports contribute to this problem but they are only a small part of this overall problem. So, an API should tackle (or enable tackling) the whole problem not just a selected portion.

**Comment by alexr00** [comment]
Proposal as of the time of this comment:
```ts
	export enum PortAutoForwardAction {
		Notify,
		OpenBrowser,
		Silent,
		Ignore
	}

	export interface PortAttributes {
		port: number;
		autoForwardAction: PortAutoForwardAction
	}

	export interface PortAttributesProvider {
		providePortAttributes(port: number, pid: number, token: CancellationToken): ProviderResult<PortAttributes>;
	}

	export namespace workspace {
		export function registerPortAttributesProvider(portRange: [number, number], provider: PortAttributesProvider): Disposable;
	}
```

**Comment by connor4312** [comment]
I can give the config approach a shot. Could scope to say 100 ports somewhere random, and try to preferentially choose those when starting to debug.

**Comment by alexr00** [comment]
@connor4312 I have merged the first pass at this API. My hope is that it is pretty easy to use, so even if it changes a lot as we continue to review it, you won't lose a lot of work by adopting it now. 

https://github.com/microsoft/vscode/blob/06044789bfaa48f87904d59cc65ee0011d698a2f/src/vs/vscode.proposed.d.ts#L2851-L2880

You'll need to implement a `PortAttributesProvider`. `providePortAttributes` will be called whenever a port is auto-detected. For any ports that shouldn't be user facing, you should return a `PortAttributes` with an `autoForwardAction: PortAutoForwardAction .Ignore`. 

If you have time to register your provider after you know the pids you care about, then you can include the `pid` your `portSelector`. 

Internally, I batch ports that have the same pid, so you should see almost no delay between calls if your ports are all in the same process. 

I'm thinking of adding a regex "commandLine" to the `portSelector` so that you can select on that too. Would that be useful for JS Debug?

**Comment by alexr00** [comment]
@connor4312 thanks for all the feedback!

> It's not clear from the existing docblock whether the port range is inclusive or exclusive
Adding to the dockblock: "The `portRange` is start inclusive and end exclusive."

> It's not clear what PortAutoForwardAction I should return if I know nothing about a port

Good point. Adding this doc: 
```
		 * Provides attributes for the given ports. For ports that your extension doesn't know about, simply don't include
		 * them in the returned array. For example, if `providePortAttributes` is called with ports [3000, 4000] but your
		 * extension doesn't know anything about those ports you can return an empty array.
```

> I noticed that providePortAttributes gets called 3x when I start a new debugging process. Checking is cheap but it's possible that this could result in a notification delay.

You should get one call per pid. Even with that, I batch the calls within the extension host, so it's only one round trip to each extension host. 

> Although I return PortAutoForwardAction.Ignore, it seems like the port is still forwarded

I tried out the vsix, and from what I can see `PortAutoForwardAction.Notify` is always getting returned. Is there something else I need to do to enable the JS Debug provider?

**Comment by alexr00** [comment]
Just to update for completeness: the JS Debug extension now uses this API successfully.

**Comment by alexr00** [comment]
Moving to April to continue discussing this proposal.

**Comment by alexr00** [comment]
Two API users so far: JS Debug and [Jupyter](https://github.com/microsoft/vscode-jupyter/blob/70073c0e1fc5a9b1b49771cde67f27c347ab2dad/src/client/common/net/portAttributeProvider.ts#L14). Let's see if we can finalize it.

**Comment by alexr00** [comment]
Moving out of this milestone since we have not specific requests to move this API to stable. Once we have a real reason to finalize we can assign to the appropriate milestone.

**Comment by andreamah** [comment]
I'm a little late to the game, but I think Live Preview would also use it to silence the notification that a port is forwarded. I have an old PR that dealt with this https://github.com/microsoft/vscode-livepreview/pull/104 (but I'd probably re-implement this if port attributes were finalized, since it's a pretty stale PR)

**Comment by alexr00** [comment]
Let's start trying to finalize this then.

**Comment by alexr00** [comment]
@brettcannon would you consider using this API in the Python extension for the Python debug adapter? Currently, when someone debugs Python with the Python extension they'll end up with some extra ports forwarded because of the debug adapter. We had the same issue with the JS Debug extension and made this API to solve the problem: https://github.com/microsoft/vscode-js-debug/blob/0d0704da3726653e2d806be528b8ab51fa952cf0/src/ui/portAttributesProvider.ts#L39-L55

**Comment by brettcannon** [comment]
> would you consider using this API in the Python extension for the Python debug adapter?

Yep, we would consider it. Did you want me to open an issue on the Python extension or did you want to?

/cc @karthiknadig

**Comment by alexr00** [comment]
@brettcannon I'll take this through another API review and then open an issue on the Python extension.

**Comment by joaomoreno** [comment]
Let's bring this back into the plan once https://github.com/microsoft/vscode-python/issues/21292 is complete.
